File: src/types/production/execution.ts
export type ExecutionStatus = 'in_progress' | 'completed' | 'aborted'

export interface FieldValue {
  field_key: string
  field_label: string
  unit: string | null
  phase: 'start' | 'end'
  value_text: string | null
  value_numeric: number | null
  value_bool: boolean | null
  is_abnormal: boolean
  remark: string | null
}

export interface EquipmentSnapshot {
  equipment_id: string
  equipment_no: string
  equipment_name: string
}

export interface Execution {
  id: string
  batch_id: string
  node_id: string
  node_name: string | null
  execution_seq: number
  status: ExecutionStatus
  owner_id: string | null
  owner_name: string | null
  started_at: string
  started_by_name: string | null
  finished_at: string | null
  finished_by_name: string | null
  is_deviation: boolean
  deviation_reason: string | null
  remark: string | null
  equipments: EquipmentSnapshot[]
  field_values: FieldValue[]
}

export interface FieldValueInput {
  field_key: string
  value: boolean | number | string | null
}

export interface StartExecutionInput {
  node_id: string
  owner_id?: string | null
  owner_name?: string | null
  equipment_ids?: string[]
  field_values?: FieldValueInput[]
  deviation_reason?: string | null
  remark?: string | null
}

export interface CompleteExecutionInput {
  field_values?: FieldValueInput[]
  remark?: string | null
}

export interface NodeExecutionListItem {
  id: string
  batch_id: string
  batch_no: string
  execution_seq: number
  status: ExecutionStatus
  owner_name: string | null
  started_at: string
  finished_at: string | null
  is_deviation: boolean
  abnormal_count: number
}

File: app/modules/production/schemas/execution.py
"""节点执行 API 契约。"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class FieldValueIn(BaseModel):
    field_key: str
    value: bool | float | str | None = None


class FieldValueOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    field_key: str
    field_label: str
    unit: str | None
    phase: str
    value_text: str | None
    value_numeric: float | None
    value_bool: bool | None
    is_abnormal: bool
    remark: str | None


class EquipmentSnapshotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    equipment_id: uuid.UUID
    equipment_no: str
    equipment_name: str


class ExecutionStartIn(BaseModel):
    node_id: uuid.UUID
    owner_id: uuid.UUID | None = None
    owner_name: str | None = Field(default=None, max_length=50)
    equipment_ids: list[uuid.UUID] = []
    field_values: list[FieldValueIn] = []
    deviation_reason: str | None = None
    remark: str | None = None


class ExecutionCompleteIn(BaseModel):
    field_values: list[FieldValueIn] = []
    remark: str | None = None


class ExecutionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    batch_id: uuid.UUID
    node_id: uuid.UUID
    node_name: str | None = None  # service 组装时填充
    execution_seq: int
    status: str
    owner_id: uuid.UUID | None
    owner_name: str | None
    started_at: datetime
    started_by_name: str | None
    finished_at: datetime | None
    finished_by_name: str | None
    is_deviation: bool
    deviation_reason: str | None
    remark: str | None
    equipments: list[EquipmentSnapshotOut] = []
    field_values: list[FieldValueOut] = []


class NodeExecutionListItem(BaseModel):
    """工序视角的执行记录行（跨批次）。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    batch_id: uuid.UUID
    batch_no: str
    execution_seq: int
    status: str
    owner_name: str | None
    started_at: datetime
    finished_at: datetime | None
    is_deviation: bool
    abnormal_count: int
